stablehlo/dialect/StablehloOps.cpp

```diff
@@ -2861,6 +2861,98 @@ LogicalResult ClampOp::reifyReturnTypeShapes(
 // ComplexOp
 //===----------------------------------------------------------------------===//
 
+namespace {
+// Utility function, used by printSelectOpType and
+// printSameOperandsAndResultType. Given a FunctionType, assign the types
+// to operands and results, erroring if any mismatch in number of operands
+// or results occurs.
+ParseResult assignFromFunctionType(OpAsmParser& parser, llvm::SMLoc loc,
+                                   ArrayRef<Type*> operands, Type& result,
+                                   FunctionType& fnType) {
+  assert(fnType);
+  if (fnType.getInputs().size() != operands.size()) {
+    return parser.emitError(loc)
+           << operands.size() << " operands present, but expected "
+           << fnType.getInputs().size();
+  }
+
+  // Set operand types to function input types
+  for (auto [operand, input] : llvm::zip(operands, fnType.getInputs())) {
+    *operand = input;
+  }
+
+  // Set result type
+  if (fnType.getResults().size() != 1) {
+    return parser.emitError(loc, "expected single output");
+  }
+  result = fnType.getResults()[0];
+
+  return success();
+}
+
+// getInferredComplexType takes a complex tensor type and returns a
+// type that maintains the shape, but removes the complex type for the
+// underlying data type
+//   Ex: tensor<4xcomplex<f32>>  -->  tensor<4xf32>
+Type getInferredComplexType(Type result) {
+  assert(result.isa<TensorType>() &&
+         result.cast<TensorType>().getElementType().isa<ComplexType>());
+  TensorType tensorTy = result.cast<TensorType>();
+  ComplexType complexTy = tensorTy.getElementType().cast<ComplexType>();
+  Type elementTy = complexTy.getElementType();
+
+  return hlo::getSameShapeTensorType(tensorTy, elementTy);
+}
```

Comment:
Consider using `createRealType` which is also defined in this file and does the same thing modulo the assert.

Reply:
Nice! I moved the function definition from the `ImagOp` section to the `ComplexOp` section and kept the code comments.